gitlabform/processors/project/job_token_scope_processor.py

```diff
@@ -0,0 +1,213 @@
+            )
+
+        existing_allowlist = job_token_scope.allowlist.list()
+
+        project_ids_to_allow = self._get_target_project_ids_from_config(
+            projects_allowlist
+        )
+
+        if len(project_ids_to_allow) > 0:
+            self._add_projects_to_allowlist(
+                project, job_token_scope, existing_allowlist, project_ids_to_allow
+            )
+            job_token_scope.refresh()
+
+        if enforce:
+            info(
+                "Enforce enabled, removing projects no longer defined in config from allowlist"
+            )
+            self._remove_projects_from_allowlist(
+                project, job_token_scope, existing_allowlist, project_ids_to_allow
+            )
+
+    def _process_groups(
+        self,
+        job_token_scope: ProjectJobTokenScope,
+        groups_allowlist: List,
+        enforce: bool,
+    ):
+        if not groups_allowlist and enforce:
+            warning(
+                "Process will remove existing groups from allowlist, as none set in configuration"
+            )
+
+        existing_allowlist = job_token_scope.groups_allowlist.list()
+
+        group_ids_to_allow = self._get_target_group_ids_from_config(groups_allowlist)
+
+        if len(group_ids_to_allow) > 0:
+            self._add_groups_to_allowlist(
+                job_token_scope, existing_allowlist, group_ids_to_allow
+            )
+            job_token_scope.refresh()
+
+        if enforce:
+            info(
+                "Enforce enabled, removing groups no longer defined in config from allowlist"
+            )
+            self._remove_groups_from_allowlist(
+                job_token_scope, existing_allowlist, group_ids_to_allow
+            )
+
+    def _remove_groups_from_allowlist(
+        self,
+        job_token_scope: ProjectJobTokenScope,
+        existing_allowlist,
+        target_group_ids: List,
+    ):
+        group_ids_to_remove = self._get_ids_to_remove_from_allowlist(
+            existing_allowlist, target_group_ids
+        )
+        for group_id in group_ids_to_remove:
+            job_token_scope.groups_allowlist.delete(group_id)
+            verbose("Deleted group %s from allowlist", group_id)
+            job_token_scope.save()
+
+    def _add_groups_to_allowlist(
+        self, job_token_scope, existing_allowlist, group_ids_to_allow
+    ):
+        group_ids_to_add = self._get_ids_to_add_to_allowlist(
+            existing_allowlist, group_ids_to_allow
+        )
+        for group_id in group_ids_to_add:
+            job_token_scope.groups_allowlist.create({"target_group_id": group_id})
+            verbose("Added group %s to allowlist", group_id)
+            job_token_scope.save()
+
+    def _add_projects_to_allowlist(
+        self, project, job_token_scope, existing_allowlist, project_ids_to_allow
+    ):
+        project_ids_to_add = self._get_ids_to_add_to_allowlist(
```

Comment:
Not sure if it's just the naming of things or if there's some optimization possible... But I find this a little hard to follow. The function already received a list of project IDs to add to allowlist. Here we're calling another function to "get IDs to add to allowlist". I was already a bit confused in the calling function why "existing allowlist" is being passed because I expected this function to just do as the name says... Add some IDs to allowlist. :)